pkg/codegen/docs/description.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package docs
+
+import (
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+const (
+	beginCodeBlock = "<!--Start PulumiCodeChooser -->"
+	endCodeBlock   = "<!--End PulumiCodeChooser -->"
+)
+
+type codeLocation struct {
+	open  int
+	close int
+}
+
+func getCodeSection(doc string) []codeLocation {
+	var fences []codeLocation
+
+	startIndex := 0
+	for {
+		open := strings.Index(doc[startIndex:], beginCodeBlock)
+		if open == -1 {
+			break
+		}
+		var fence codeLocation
+		fence.open = startIndex + open
+		startIndex += open + len(beginCodeBlock)
+		closing := strings.Index(doc[startIndex:], endCodeBlock)
+
+		contract.Assertf(closing != -1, "this should never happen: "+
```

Comment:
hmm not sure the best way to handle this...  but I do lean towards returning an error here and having the error bubble back up to the [gen call](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/docs/gen.go#L1828) which will eventually return back to `resourcedocsgen`. Then we could let `resourcedocsgen` decide how it wants to handle it since it is ultimately the one who is calling this code.

Though, I believe as it stands, if `resourcedocsgen` encounters any errors it would fail and exit the process, so the end result would be the docs generation failing anyway. Given that it would fail anyway on the `resourcedocsgen` end,  I'm not super worried about this behavior and don't want to create too much additional work here.